Add accounts blocklisting design doc (#6174)
hashgraph/hedera-services: hedera-node/docs/design/services/smart-contract-service/accounts-blocklisting.md

In both cases, this results in the private key losing control over the EVM address derived from the corresponding public key.

This means that there is a difference in how some situations, like accidental transfers to known EVM addresses, will be handled in Ethereum and in Hedera:
- In Ethereum, everyone that knows the corresponding private key will be able to authorize a transaction from the known address, and the first to send a transaction would be eligible to withdraw the funds.
- In Hedera, only the account with the associated corresponding alias would be eligible to withdraw the funds. If another key controls this account, then only that key can be used to withdraw the funds.

## Goals

- We want to block a list of known addresses, so that no one can “squat” on them and exercise full control on any accidentally sent and blocked funds. Sending funds to such blocked addresses should not be successful as these accounts are considered “compromised”.

## Non Goals

- We don't consider unblocking already blocked accounts.
- We don't consider what happens to a blocked account when it expires.

## Architecture

> The implementation related to blocked accounts will be gated by a feature flag.

Reviewer: Maybe we should list the feature flags and their effect, the structure of the file used to list all the private keys

Author: There is only one feature flag related to this, do you mean to also add the property that has the name of the resource file that contains the blocklist as well?

As for the structure of the file, there is a `File containing the list of blocked accounts` two lines below this which describes that, do you mean to update it in some way?

Reviewer: Somehow I missed the next section when I wrote this, my bad 😊

# Accounts Blocklisting

## Purpose

In Ethereum, everyone that knows a private key is able to authorize a transaction from the corresponding EVM public address, and the first to send a transaction would be eligible to withdraw any funds from that address.

> In Hedera, it is possible for an account with associated EVM address alias to have a key that does not correspond to the alias. This is what we refer to as “EVM address squatting” i.e. using a key that’s different from the key that’s used to derive the EVM address alias from.

Reviewer 2: I wouldn't term this as such given it's a likely occurrence of safe key practices i.e. key rotation.
I think the squatting here is the act of creating account with matching evm address to publicly known exposed private keys - that's the act of squatting.

Author: Updated.

It is possible to “squat” on an EVM address in a couple of ways:
1. Create an account with an alias and then update the account key.
2. Create an account with an alias and a key that’s different from the key that’s used to derive the alias from.

> In both cases, this results in the private key losing control over the EVM address derived from the corresponding public key.

Reviewer 3: which "private key" has lost "control"?  I think this means the EVM address' private key is no longer in _sole_ control - that there are now _two_ keys that can drain the address (this is the "first to send a transaction" in the previous sentence but with _two_ keys available instead of just one (shared) one.

Reviewer: Not really, the "EVM address' private key" can no longer be used to control the account if the account has a different key - only the key set on the account can be used to sign transactions that transfer funds out of that account. 
Only if the account is updated with a 1 of 2 threshold key with (1) the new account key and (2) "EVM address' private key", then both keys could be used to control the account.

This means that there is a difference in how some situations, like accidental transfers to known EVM addresses, will be handled in Ethereum and in Hedera:
- In Ethereum, everyone that knows the corresponding private key will be able to authorize a transaction from the known address, and the first to send a transaction would be eligible to withdraw the funds.

> - In Hedera, only the account with the associated corresponding alias would be eligible to withdraw the funds. If another key controls this account, then only that key can be used to withdraw the funds.

Reviewer 3: Ah! This is pretty strong.  Stronger than I suggested above.  In this case not only is the Hedera account key holder able to drain the account but the Ethereum private key holder is _unable_ to.  Correct?

Reviewer: correct. that's how Hedera differs in behavior, since in Ethereum it's not _possible_ to change the controlling key of an acocunt, it's always the key used to derive the EVM address from.

- As those default private test keys are well-known, anyone can use them to withdraw the funds from the account, so in effect, the funds are lost.

In Hedera (current behavior):
- While executing tests against different environments, using some default private test keys (e.g. the ones configured with Hedera Local Node) to create and interact with accounts, I accidentally executed my tests on Mainnet and transferred some funds to an account with alias derived from one of the Hedera Local Node keys.
- As those default private test keys are well-known, anyone could have used them to create an account with the same alias but with different account admin key, leaving me with no control over the funds.

In Hedera (desired behavior):
- While executing tests against different environments, using some default private test keys (e.g. the ones configured with Hedera Local Node) to create and interact with accounts, I accidentally executed my tests on Mainnet and transferred some funds to an account with alias derived from one of the Hedera Local Node keys.

> - As those default private test keys are well-known, the ledger can be configured to own those EVM addresses and block any transfers to them, so that I cannot accidentally send funds to them.

Reviewer 2: Is the inadvertent effects of this that you can't send to any of the test accounts for some of these popular products when run locally?
For production envs this is right, for local node for example this feature might have to be disabled.

Reviewer: Yes, this is correct! The accounts should not be created on local environments because the test accounts with well-known keys and addresses become unusable!

## Goals

- We want to block a list of known addresses, so that no one can “squat” on them and exercise full control on any accidentally sent and blocked funds. Sending funds to such blocked addresses should not be successful as these accounts are considered “compromised”.

## Non Goals

- We don't consider unblocking already blocked accounts.
- We don't consider what happens to a blocked account when it expires.

## Architecture

The implementation related to blocked accounts will be gated by a feature flag.

### File containing the list of blocked accounts

> We will store the list of blocked accounts in a regular CSV resource file.

Reviewer 3: With this approach we can update the blocked accounts list at every release (if/as necessary).  Is it possible we may want to update it more often, or just between releases?  In that case a proper Hedera file could be used?  (I believe that's used for other properties already.)

(If this has been considered and rejected possibly a sentence could be added to the doc about "alternative approaches considered" - I've seen that in other design docs.)

Reviewer: Hmm, we may consider this. What are the advantages of using a "proper Hedera file"? 
I think one of the main advantages of using the Hedera File Service is that updates to the file automatically take effect without needing to restart or upgrade the network, but this is not applicable to our case because we create the accounts on upgrade and perform a migration for the mirror nodes.
One more thing to consider is that removing a private key from the file will **not** uncreate the account, so basically we can add more private keys to the list, but not remove already created ones - or more precisely, we can remove already created ones, but that will only remove them from the file, not from the network!

Reviewer 3: I should have worked out those implications, thanks!

Reviewer 3: But, you know, in case someone _else_ doesn't work out those implications ... perhaps a section at the bottom of the doc for "other approaches considered" could have this. It would be a good place to remind readers that putting an address on this blocklist is not reversible!

Author: Added a section a the end of the doc for this.

### Creating the blocked accounts in state
After the node starts (on handling the first transaction), for each of the EVM addresses that we want to block create a regular account with the following properties:
- `receiverSigRequired` equal to `true`, thus blocking all future transfers to these accounts
- `key` equal to the `GENESIS` key
- `alias` equal to the EVM address derived from the corresponding private key
- `memo` if specified for the blocked EVM address in the resource file, indicating the reason for blocking the account

Anyone with access to the `GENESIS` key will have sole control on the accounts and the fact that a given alias can be associated only to a single account prevents squatting on the EVM addresses.
The EVM already has checks integrated for accounts configured with `receiverSigRequired`.
This approach does not require us to introduce a property file with random addresses and to have custom checks and logic executed in the EVM.

### Externalizing the blocked accounts
All created blocked accounts will be externalized as synthetic account creations.

> ### New Classes

Reviewer 2: Should add an acceptance tests section to highlight acceptance  test case cases for successful implementation

Author: Added.

# Accounts Blocklisting

## Purpose

In Ethereum, everyone that knows a private key is able to authorize a transaction from the corresponding EVM public address, and the first to send a transaction would be eligible to withdraw any funds from that address.

> In Hedera, it is possible for an account with associated EVM address alias to have a key that does not correspond to the alias. When an account is created with EVM address alias matching some publicly known exposed private key, this is what we refer to as “EVM address squatting”.

Reviewer 2: I think this is closer to the general essence. The previous wording highlight one case. If the keys don't differ then it's just a normal case of account creation that claims the noted address.
```suggestion
In Hedera, it is possible for an account with associated EVM address alias to have a key that does not correspond to the alias. When an account is created with EVM address alias but has a non matching account key, this is what we refer to as “EVM address squatting”.
```

Author: Updated.

> In Hedera, it is possible for an account with associated EVM address alias to have a key that does not correspond to the alias. When an account is created with EVM address alias matching some publicly known exposed private key, this is what we refer to as “EVM address squatting”.

Reviewer 2: Should we use "reserve" instead of "squat"?
Squat may have a negative connotation.

Author: Replaced all occurrences of "squat" with "reserve".

It is possible to “squat” on an EVM address in a couple of ways:
1. Create an account with an alias and then update the account key.
2. Create an account with an alias and a key that’s different from the key that’s used to derive the alias from.

> In both cases, this results in the private key losing control over the EVM address derived from the corresponding public key.

Reviewer 2: ```suggestion
In both cases, this results in the private key losing control over the account with EVM address derived from the corresponding private-public key pair.
```

Author: Updated.

This means that there is a difference in how some situations, like accidental transfers to known EVM addresses, will be handled in Ethereum and in Hedera:
- In Ethereum, everyone that knows the corresponding private key will be able to authorize a transaction from the known address, and the first to send a transaction would be eligible to withdraw the funds.

> - In Hedera, only the account with the associated corresponding alias would be eligible to withdraw the funds. If another key controls this account, then only that key can be used to withdraw the funds.

Reviewer 2: Rewording suggestion
```suggestion
- In Hedera, only the key that controls an account can be used to withdraw the funds. This key may not correspond to the account alias.
```

Author: Updated.

- `receiverSigRequired` equal to `true`, thus blocking all future transfers to these accounts
- `key` equal to the `GENESIS` key
- `alias` equal to the EVM address derived from the corresponding private key
- `memo` if specified for the blocked EVM address in the resource file, indicating the reason for blocking the account

Anyone with access to the `GENESIS` key will have sole control on the accounts and the fact that a given alias can be associated only to a single account prevents reserving of the EVM addresses.
The EVM already has checks integrated for accounts configured with `receiverSigRequired`.
This approach does not require us to introduce a property file with random addresses and to have custom checks and logic executed in the EVM.

### Externalizing the blocked accounts
All created blocked accounts will be externalized as synthetic account creations.

### New Classes
`BlocklistAccountCreator` class will encapsulate the logic for reading blocked accounts from file and creating them in state.

## Acceptance Tests

> * Verify that blocked accounts are created in state and that synthetic records are externalized for them when the node starts after the first transaction is handled.

Author: Added. Thanks!
